app/controllers/whats_app/three_sixty_dialog_webhook_controller.rb

```diff
@@ -0,0 +1,99 @@
+        handle_subscribe_contributor(contributor)
+      end
+
+      adapter.consume(message_params.to_h) { |message| message.contributor.reply(adapter) }
+    end
+    # rubocop:enable Metrics/AbcSize
+
+    def create_api_key
+      channel_ids = create_api_key_params[:channels].split('[').last.split(']').last.split(',')
+      client_id = create_api_key_params[:client]
+      Setting.three_sixty_dialog_client_id = client_id
+      channel_ids.each do |channel_id|
+        WhatsAppAdapter::CreateApiKey.perform_later(channel_id: channel_id)
+      end
+      render 'onboarding/success'
+    end
+
+    private
+
+    def message_params
+      params.permit({ three_sixty_dialog_webhook: [contacts: [:wa_id, { profile: [:name] }],
+                                                   messages: [:from, :id, :type, :timestamp, { text: [:body] }, { context: %i[from id] },
+                                                              { button: [:text] }, { image: %i[id mime_type sha256] },
+                                                              { voice: %i[id mime_type sha256] },
+                                                              { video: %i[id mime_type sha256] },
+                                                              { audio: %i[id mime_type sha256] },
+                                                              { errors: %i[code details title] },
+                                                              { document: %i[filename id mime_type sha256] },
+                                                              { location: %i[latitude longitude] },
+                                                              { contacts: [{ org: {} }, { addresses: [] }, { emails: [] }, { ims: [] },
+                                                                           { phones: %i[phone type wa_id] }, { urls: [] },
+                                                                           { name: %i[first_name formatted_name last_name] }] }]] },
+                    contacts: [:wa_id, { profile: [:name] }],
+                    messages: [:from, :id, :type, :timestamp, { text: [:body] }, { context: %i[from id] }, { button: [:text] },
+                               { image: %i[id mime_type sha256] }, { voice: %i[id mime_type sha256] },
+                               { video: %i[id mime_type sha256] },
+                               { audio: %i[id mime_type sha256] },
+                               { errors: %i[code details title] },
+                               { document: %i[filename id mime_type sha256] },
+                               { location: %i[latitude longitude] },
+                               { contacts: [{ org: {} }, { addresses: [] }, { emails: [] }, { ims: [] },
+                                            { phones: %i[phone type wa_id] }, { urls: [] },
+                                            { name: %i[first_name formatted_name last_name] }] }])
+    end
+
+    def create_api_key_params
```

Comment:
Could we maybe use a different controller for messaging and setup?

Reply:
Yeah, originally I added this in a separate controller, see [`9468a86` (#1645)](https://github.com/tactilenews/100eyes/pull/1645/commits/9468a8650a33ac58c61b6133365a81fd6e450c27)

I put it here because I wanted to differentiate between client/contributor onboarding, but maybe we could separate it again and rename it to `WhatsApp::SetupController`, or something similar?